### ISSUE TITLE
Saves, Presets, Defaults

### ISSUE THREAD

**Issue by ZenRevision** [gileli121/WindowTop#371]
Just trying WindowTop for first time. It is almost precisely what I was looking for, so I'll definitely be purchasing a license. The are just a couple stumbling blocks though.

There should be a list of saved window configs, that we can edit or delete without reopening a given window. We should also be able to manually add saved configs by inputting a window name, application.exe, or guid, and edit which settings get applied.

We should also be able to create presets, that can be activated with hotkeys or via the menu. So a _save preset_ hotkey and/or when we open the menu, there could be a _save preset_ icon next to the save icon. The save preset would just save all the settings currently active, but not targeted to a specific window. When saving a preset it should open a small dialogue to give the preset a name, and optionally assign a hotkey.

Then the hotkey assigned to a preset would activate all associated settings, and in the menu there could be a preset button, that would open a dropdown list from which one could activate saved presets. Also, in settings, along with a list of saved window configs (mentioned above), there should be a list of saved presets, where we could edit, delete, or manually add new presets.

Finally, it seems like we should be able to edit defaults. So when I activate the Glass effect, for instance, it defaults to whatever I set for the default Glass settings (for example, 75% Transparent, 50% Blur, 50% Image, 50% Text, with Filter Images on). This could just be a _remember settings_ tickbox in the menu dropdowns, or even just an option in settings to _remember all settings_ globally. Edit: I guess we can set some defaults, but not all. It seems like _Glass_ doesn't use the global default transparency, and I don't see anywhere to set default _Glass_ settings at all.

I will purchase a license regardless, but if editable presets, and/or editable defaults (edit; or rather, more comprehensive defaults), where implemented, I would be able to recommend this software whole heartedly to numerous colleagues.

As documented there are artifacts and inconsistencies with the Glass effect (I'm using Win 10, and notice artifacts in Notepad++, while Sublime Text looks almost perfect, but in Sublime Text the effect just turns off when switching tabs, or on other random interactions). I realize the issues with Glass may not be easy to fix, but I hope it can be addressed eventually as this is the only app I have found which can do transparent backgrounds with blur almost universally, which is a truly **killer feature**. One thing I think may help is to employ some antialiasing between transparent and opaque elements (text generally looks good, so maybe somehow improving the identification of UI elements for antialiasing and to try to separate them from the background more consistently).

Thanks for building such an awesome app! Looking forward to seeing how you continue developing it!

**Comment by gileli121** [comment]
@ZenRevision 
Thanks for your great feedback!

About option to modify list of saved window configuration, I am aware that it is not available. I actually even started to develop this option under another development branch but suspended to develop it due to some priorities.
But this feature to modify window configurations from list (instead of editing the settings.json file) is almost finished. I just need time to finish with implementation and release a beta for testing. I will consider if this should be the highest priority.

About the presets, can you give an example of a possible workflow with a real application?
I need an actual example to understand the suggestion better.

If you have feedback on the Glass settings, I'd appreciate it if you could open it as a separate issue. This will help us keep track of all the different aspects of the software we need to improve.

**Comment by ZenRevision** [comment]
Yeah, it's great that it sounds like the ability to modify the list of saved window configs will be implemented at some point!

It would be great if we could edit saved config's window identification method. For example, I save a config with Notepad++, then I edit the saved config to identify the Notepad++ window based on the window title, rather than the .exe, because Notepad++ includes the document name in the window title. That way I could have a saved window config that only activated when a Specific document was open in Notepad++. Ideally we could have some _target matching_ options like <contains>, <precise match>, and <doesn't contain> and _target type_ options like, <window title>, <.exe name>, and <guid>. Ideally, multiple sets of these criteria could be combined as well.

So based on that, here are a few example window matching rules.

- Example 1: 
```Apply this config if <window title> <contains> <example string>```
- Example 2: 
```Apply this config if <window title> <precisely matches> <example string>```
- Example 3: 
```Apply this config if <.exe name> <precisely matches> <example.exe> & <window title> <doesn't contain> <example string>```

I don't really think that accessing saved configs, to add/edit/delete them, really even needs a UI. This feature could just be for advanced users. You could just save these configs in .json or .xml files in a directory, using human readable file names like name_of_exe_000.json, making them easy enough to identify by name, and these file names would really just be for reference anyway, as the window targets would be determined by values within these .json files, not their file names. So users could rename these files according to their own preferences.

That way we could create new configs easily, edit existing configs easily, and delete unneeded configs easily, without the need for any major UI refactoring.

Regarding the presets I mentioned previously, as well as the Glass feature, I'll open separate issues.

**Comment by gileli121** [comment]
@ZenRevision 
Your idea to apply saved config, not by the process name, was something I had thought of before. It seems like a good idea.
I did not consider the option of not providing it with a detailed UI (so the user will, for the most part, use/edit XML/JSON files).
I agree that it may be possible not to provide a detailed UI for it because very advanced users probably use it.

About the presets, see my comment there.
It sounds like you're trying to solve the same problem and suggest two ideas (presets and this one). In that case, it's better to understand what the best proposal is here.

I think that this proposal is probably a better way to handle your need (assuming that presets also aim to handle the same need)

**Comment by ZenRevision** [comment]
Hmm, I guess I needed to use code tags. It seems my examples above where broken upon posting. Fixed now. 

The presets concept is much more related to defaults than saves, because the idea behind presets is that they wouldn't target an app/window specifically nor be applied automatically.

However, preset functionality could, I think, easily be rolled into this save feature, by just including a way to define a hotkey within the save .json/.xml and not requiring a window target (maybe including a preset key).

For example:
```
Save_Name = Example_Name
Preset = True
Window_Target = Undefined
Hotkey = "0x10"+"0x12"+"0x77" ; Shift+Alt+F8

; Configuration details below...
```
That way I could save a window config for any window. Go into the saved configs directory and edit that saved file. Delete the window target values (or maybe change the target value to ```<Undefined>```) & add a hotkey. Then I could toggle that config on/off for any window with a hotkey.

I do actually think that presets would be useful to less advanced users and thus benefit from a UI as described in the other issue. However, as an initial implementation it seems like including presets as as I've just described here without any UI and as part of the advanced save editing would be awesome. Then at some point in the future it may be worth thinking about adding UI elements, at which point the underlying functionality would already be in place.

**Comment by gileli121** [comment]
@ZenRevision 

According to what you wrote here:

> The presets concept is much more related to defaults than saves, because the idea behind presets is that they wouldn't target an app/window specifically nor be applied automatically.


It seems that, based on your initial thought, you prefer the save config per app or [window](https://github.com/WindowTop/WindowTop-App/issues/371#issuecomment-2342711857).
But then you rolled it out because of the issue that it applies every time (when the window opens), and you don't want this specific behavior (you only wanted to save the config but only sometimes apply it). It seems like a good idea.
So, going to far with a complete different concept? what if I only develop [this proposal](https://github.com/WindowTop/WindowTop-App/issues/371#issuecomment-2342711857) *and also option to not auto apply the config*? so you keep using the existing save config feature. but it will include a button & hotkey to apply this config (if detected for that window by process name or something like what you described [here](https://github.com/WindowTop/WindowTop-App/issues/371#issuecomment-2342711857)) .


The only benefit of the preset concept is that you can switch between configurations for every window very quickly.
But do you really see a real use case for why you want to do it? For me, it looks like you only want one configuration for a window or app but without the part that applies when that window opens.


> I do actually think that presets would be useful to less advanced users and thus benefit from a UI as described in the other issue.

I want WindowTop to be simple—that is, all its features should be simple and powerful at the same time. Otherwise, it will be like PowerToys. 
This is also another argument for why presets are not the best idea, mainly if enhancements (as I said above) to the save-config feature can serve the same need.

But I don't want you to think that I argue with you. In some way, the users are always right. If multiple users see a need for presets, I will develop this feature. And I mainly do it to ensure that you are 100% sure that you ask for the right idea.

**Comment by gileli121** [comment]
> As documented there are artifacts and inconsistencies with the Glass effect (I'm using Win 10, and notice artifacts in Notepad++, while Sublime Text looks almost perfect

If you have NVIDIA GPU, it supports [CUDA](https://en.wikipedia.org/wiki/CUDA). it means having a code that runs directly inside the GPU (instead of CPU). if you enable this option:
![image](https://github.com/user-attachments/assets/1a50d493-263a-478c-a5aa-33736d6835a3)

The performance will be 4x times faster (according to my computer), and because it is not exactly the same code (different code needed for GPU), it may also give better results. I also tested it more with CUDA.
Maybe it worth for you trying it in case it is relevant for you. It should work better for sure 

See how it looks on my end:
![image](https://github.com/user-attachments/assets/bcd547d1-0cc1-4fdb-b5ae-d7bea5649b18)

**Comment by gileli121** [comment]
Sorry for the confusion. I deleted my last two comments here and answered you [here](https://github.com/WindowTop/WindowTop-App/issues/375#issuecomment-2350658618) instead.
